### PR TITLE
Add a RouteBuilderLifecycleStrategy to intercept RouteBuilder related events

### PR DESCRIPTION
**Pull request by lburgazzoli**

### CODE REVIEW

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.builder;
+
+public interface RouteBuilderLifecycleStrategy {
```
Comment (davsclaus) [inline core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: Move to camel-api and add javdoc to class level

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java [hunk 2]
```diff
@@ -478,12 +493,21 @@ protected void checkInitialized() throws Exception {
             if (camelContext.adapt(ExtendedCamelContext.class).getErrorHandlerFactory() instanceof ErrorHandlerBuilder) {
                 setErrorHandlerBuilder((ErrorHandlerBuilder)camelContext.adapt(ExtendedCamelContext.class).getErrorHandlerFactory());
             }
+
+            for (RouteBuilderLifecycleStrategy interceptor: lifecycleInterceptors) {
```
Comment (davsclaus) [inline core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: Fix CS

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java [hunk 3]
```diff
@@ -478,12 +493,21 @@ protected void checkInitialized() throws Exception {
...
+            }
+
             configure();
             // mark all route definitions as custom prepared because
             // a route builder prepares the route definitions correctly already
             for (RouteDefinition route : getRouteCollection().getRoutes()) {
                 route.markPrepared();
             }
+
+            for (RouteBuilderLifecycleStrategy interceptor: lifecycleInterceptors) {
```
Comment (davsclaus) [inline core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: Fix CS

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java [hunk 4]
```diff
@@ -71,6 +71,7 @@
     private List<TransformerBuilder> transformerBuilders = new ArrayList<>();
     private List<ValidatorBuilder> validatorBuilders = new ArrayList<>();
     private RoutesDefinition routeCollection = new RoutesDefinition();
+    private List<RouteBuilderLifecycleStrategy> lifecycleInterceptors = new ArrayList<>();
```
Comment (davsclaus) [inline core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: Make final

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * This method is invoked before the {@link RouteBuilder#configure()}method is invoked.
```
Comment (davsclaus) [inline core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: Fix spacing in javadoc, also for the other method

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java (hunk 1 above)
Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: even if it is RouteBuilder specific ?

Comment (davsclaus) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: Ah sorry yeah thats not doable, so it should be here. Its just that its an api that is now a bit hidden inside core-engine.

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java (hunk 2 above)
Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: done

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java (hunk 3 above)
Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: done

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java (hunk 4 above)
Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilder.java]: done

File: core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java (hunk 5 above)
Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: done

Comment (lburgazzoli) [reply core/camel-core-engine/src/main/java/org/apache/camel/builder/RouteBuilderLifecycleStrategy.java]: done

**Pull request merged**